src/main/scala/chisel3/util/BoringUtils.scala

```diff
@@ -0,0 +1,59 @@
+// See LICENSE for license details.
+
+package chisel3.util
+
+import chisel3._
+import chisel3.experimental._
+import chisel3.internal.{InstanceId, NamedComponent}
+import firrtl.transforms.DontTouchAnnotation
+import firrtl.passes.wiring.{WiringTransform, SourceAnnotation, SinkAnnotation}
+
+/** Utilities for generating synthesizeable cross module references.
+  *
+  * @example {{{
+  * class ModuleA extends Module with BoringUtils {
+  *   val a = Reg(Bool())
+  *   addSource(a, "unique_identifier")
+  * }
+  * class ModuleB extends Module with BoringUtils {
+  *   val b = Wire(Bool())
+  *   addSink(b, "unique_identifier")
+  * }
+  * class ModuleC extends Module with BoringUtils {
+  *   val c = Wire(Bool())
+  *   addSink(c, "unique_identifier")
+  * }
+  * }}}
+  */
+trait BoringUtils extends BaseModule {
+  self: BaseModule =>
+
+  /** Add a named source cross module reference
+    *
+    * @param component source circuit component
+    * @param name unique identifier for this source
+    */
+  def addSource(component: NamedComponent, name: String): Unit = {
```

Comment:
I've always thought it would be cool if such an API could be a little less Stringly-typed. For example:
```scala
object bore {
  def apply(source: Data, sinks: Data*): Unit = {
    def genName: = source. instanceName // delayed just like .toNamed
    annotate(new ChiselAnnotation with RunFirrtlTransform {
      def toFirrtl = SourceAnnotation(source.toNamed, genName)
      def transformClass = classOf[WiringTransform]
    })
    for (sink <- sinks) {
      annotate(new ChiselAnnotation with RunFirrtlTransform {
        def toFirrtl = SinkAnnotation(sink.toNamed, genName)
        def transformClass = classOf[WiringTransform]
      })
    }    
  }
}
```

Reply:
I'm not really sure how to deal with the fact that the names need to be distinct though... There could be a thread-safe datastructure that keeps track? The main issue with that is there is no way to clear it after Chisel elaboration finishes so multiple Chisel invocations could impact each other through that structure. Perhaps Chisel could provide an `endOfElaboration` hook...

Reply:
The stringly typed nature is annoying, I agree. That was entirely motivated by the inability of (in my understanding) Chisel to support arbitrary cross module references. You'd like to be able to describe, "Connect this source component to this sink component," with a single method. To do that, you need some way of being able to specify a component outside of the current module scope. Otherwise, it appears to me like the annotator sketch above would be limited to doing connections in one module scope (or, provide redundancy with `:=`). Maybe I'm missing something... or some way of talking about specific `Data` objects.

Is there a known way to do this?

Also, the parenthetical above may motivate an actual "bore connector", e.g., `:=>`, `:=b=`, `:=>=`.

I like this annotator and I'm amenable to a singleton managing state. Something like Rocket's elaboration artefacts (and it's British spelling) seems to make sense as a model for how to add hooks.

Reply:
Perhaps I am misunderstanding, but with typical Chisel code there's no reason why you *can't* express a cross-module reference. For example:
```scala
class Bar extends Module {
  val io = IO(new Bundle {})
  val myReg = RegInit(8.U)
}
class Foo extends Module {
  val io = IO(new Bundle {})
  val bar = Module(new Bar)
}
class Top extends Module {
  val io = IO(new Bundle { val out = Output(UInt(8.W)) })
  val foo = Module(new Foo)
  bore(foo.bar.myReg, io.out)
}
```
Now of course, you can't access fields if inner scopes or functions, but anything that makes it to a public val can be referenced, which due to reflection-based naming, is obviously how much Chisel code looks!

An obviously if you use `foo.bar.myReg` in `Top` in any *Chisel* operations, it will error, but you can certainly do such a cross-module annotation for the purposes of BoringUtils

Reply:
You're right! I don't know why that didn't occur to me. All the `val`s are public. Let me think about this... This pattern would make it so that the boring has to happen in the lowest common ancestor module or higher. I've always assumed that you'd want to annotate at the component. Cross module references wouldn't work there, but there may be a non-stringly typed way...